R/reindent.R

```diff
@@ -0,0 +1,134 @@
+#'   terms of col2. We need to take the last from the first child because calls
+#'   like package::function() can have three elements.
+#' @examples
+#' \dontrun{
+#' # not re-indented
+#' call(call(
+#'   xyz
+#' ))
+#' # re-indented
+#' call(call(1,
+#'           2))
+#' }
+#' @importFrom purrr map_lgl
+update_indention_ref_fun_call <- function(pd_nested) {
+  current_is_call <- pd_nested$token_before[2] %in% c("SYMBOL_FUNCTION_CALL")
+  non_comment <- which(pd_nested$token != "COMMENT")
+  first_non_comment_after_call <- non_comment[non_comment > 2][1]
+  if ((current_is_call) &&
+      nrow(pd_nested) > 3 &&
+      pd_nested$lag_newlines[first_non_comment_after_call] == 0) {
+    candidates <- 3:(nrow(pd_nested) - 1)
+
+    child_is_call <- map_lgl(pd_nested$child, is_function_call)
+    child_is_curly_expr <- map_lgl(pd_nested$child, is_curly_expr)
+    child_is_on_same_line <- cumsum(pd_nested$lag_newlines) == 0
+    call_on_same_line <- child_is_call & child_is_on_same_line
+    to_indent <- setdiff(candidates, which(call_on_same_line | child_is_curly_expr))
+
+    pd_nested$indention_ref_id[to_indent] <- last(pd_nested$child[[1]]$id)
+  }
+  pd_nested
+}
+
+#' @describeIn update_indention_ref Updates the reference id for all
+#'   tokens in `pd_nested` if `pd_nested` contains a function declaration.
+#'   Tokens inside a function declaration are are re-indented,
+#'   that is, they are indented up to the level at which the token FUNCTION
+#'   ends in terms of col2.
+#' @examples
+#' \dontrun{
+#' a <- function(x,
+#'               y) {
+#' x + y
+#' }
+#' }
+update_indention_ref_fun_dec <- function(pd_nested) {
+  if (pd_nested$token[1] == c("FUNCTION") &&
+      nrow(pd_nested) > 3) {
+    seq <- 3:(nrow(pd_nested) - 1)
+    pd_nested$indention_ref_id[seq] <- pd_nested$id[1]
+  }
+  pd_nested
+}
+
+#' Check whether a parse table corresponds to a a certain expression
+#'
+#' @param pd A parse table.
+#' @name pd_is
+NULL
+
+#' @describeIn pd_is Checks whether `pd` contains an expression wrapped in
+#'   curley brackets.
+is_curly_expr <- function(pd) {
+  if (is.null(pd)) return(FALSE)
+  pd$token[1] == "'{'"
+}
+
+#' @describeIn pd_is Checks whether `pd` is a function call.
+is_function_call <- function(pd) {
+  if (is.null(pd)) return(FALSE)
+  if (is.na(pd$token_before[2])) return(FALSE)
+  pd$token_before[2] == "SYMBOL_FUNCTION_CALL"
+}
+
+
+#' Apply reference indention to tokens
+#'
+#' Applies the reference indention created with functions
+#' [update_indention_ref()] to the flattened parse table. The indention
+#' is applied to all token that inherit from a reference token sequentially,
+#' i.e. by looping over the target tokens.
+#' @inheritParams apply_ref_indention_one
+apply_ref_indention <- function(flattened_pd) {
+  target_tokens <- which(flattened_pd$id %in% flattened_pd$indention_ref_id)
+
+  # needs sequential update
+  for (target_token in target_tokens) {
```

Comment:
I think it needs to be sequential, because some re-indention is based on token that were itself re-indented (at least in the case of function call), so I guess `Reduce()` (or `purrr::reduce()` is the way to go.